task.. done O_O

MykytaMusaiev opened the pull request:
[DEMO LINK](https://MykytaMusaiev.github.io/js_2048_game/)

MykytaMusaiev committed:
task.. done O_O

MykytaMusaiev committed:
minor fixes

MykytaMusaiev committed:
more minor fixes

MykytaMusaiev committed:
added mobile support

MykytaMusaiev committed:
fixed win condition; add keep play button

MykytaMusaiev committed:
fixed infinitive win message; fixed button style;

MykytaMusaiev committed:
fixed hidden class

alexander-ignatow commented on src/modules/Game.class.js:
```diff
@@ -21,24 +21,176 @@ class Game {
+      if (i + 1 < slidRow.length && slidRow[i] === slidRow[i + 1]) {
+        const mergedValue = slidRow[i] * 2;
+
+        if (mergedValue === 2048) {
```
2048 is a magic number

alexander-ignatow commented on src/modules/Game.class.js:
```diff
@@ -21,24 +21,176 @@ class Game {
+
+  _checkForGameOver() {
+    for (let i = 0; i < 4; i++) {
+      for (let j = 0; j < 4; j++) {
```
4 is a magic number

even more - yes, classic 2048 is played on a 4x4 field, but  it makes sense to make this some sort of constant, not inline value (for instance - don't create board just explicitly as 4x4 matrix, rather make a change to create an NxN array, where N is provided externally)

alexander-ignatow commented on src/modules/Game.class.js:
```diff
@@ -21,24 +21,176 @@ class Game {
+
+        if (j < 3 && this.board[i][j] === this.board[i][j + 1]) {
+          return;
+        }
```
3 looks like a magic number too

why the condition j<3 is so special?

alexander-ignatow commented on src/modules/Game.class.js:
```diff
@@ -21,24 +21,176 @@ class Game {
-    // eslint-disable-next-line no-console
-    console.log(initialState);
+    this.score = 0;
+    this.status = 'idle'; // One of: 'idle', 'playing', 'win', 'lose'
```
That is why you should not use JS for a modern real world tasks
status should be an enum (TS has support for enums and for union types)

Still, this is error prone to have a status like this - with simple literals

Step 1: add following code 

```
const Status = {
Idle: 'idle',
Playing: 'playing',
// ...
}
```

Step 2 

```suggestion
    this.status = Status.Idle;
```

See more on enums in JS: https://www.sohamkamani.com/javascript/enums/

alexander-ignatow replied:
Step 3 - never compare the status with a literal directly - always use your object

`if (this.status === Status.Playing) ... `

alexander-ignatow commented on src/modules/Game.class.js:
```diff
@@ -21,24 +21,176 @@ class Game {
+      }
+    }
+
+    while (mergedRow.length < 4) {
```
the 4 looks as a magic number again

alexander-ignatow commented on src/modules/Game.class.js:
```diff
@@ -21,24 +21,176 @@ class Game {
+
+    const oldBoard = JSON.stringify(this.board);
+
+    switch (direction) {
```
direction should be an enum too - see the Status

alexander-ignatow commented on src/modules/Game.class.js:
```diff
@@ -21,24 +21,176 @@ class Game {
+        break;
+    }
+
+    const moved = JSON.stringify(this.board) !== oldBoard;
```
That JSON trick looks like a hack to me

First - what exaclty do you gain with this optimisation?
Does it happen often, when the move results in no changes on the board?

If you still deside you need this, You should change this logic - maybe craft some sort of hash function for the state? Or just copy the internal state and compare it with the old?

alexander-ignatow replied:
BTW - are you sure that you do save processor time creating and comparing JSON strings VS doing an "empty rerender"?

alexander-ignatow commented on src/modules/Game.class.js:
```diff
@@ -21,24 +21,176 @@ class Game {
-  getState() {}
+  getState() {
+    return this.board;
+  }
```
Change this to getters
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get

alexander-ignatow commented on src/modules/Game.class.js:
```diff
@@ -50,19 +202,38 @@ class Game {
-  getStatus() {}
+  getStatus() {
+    return this.status;
+  }
```
use a getter here too

alexander-ignatow commented on src/modules/Game.class.js:
```diff
@@ -50,19 +202,38 @@ class Game {
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+    ];
```
I see this zeroes repeatedly.
Maybe you can add some function - like 

```
generateStartField = () => [
      [0, 0, 0, 0],
      [0, 0, 0, 0],
      [0, 0, 0, 0],
      [0, 0, 0, 0],
    ]
```

And use it anywhere you need this starting value

alexander-ignatow reviewed (changes requested):
Good, but lets make the code better

One more suggestion - it is out of the scope, but if I will just open the demo without reading the task, I cannot simply play the game
It would be nice to add some instructions bellow (like use arrow keys, etc)

MykytaMusaiev committed:
big update and some fixes

MykytaMusaiev commented:
@alexander-ignatow HI!
Changed all  "magic numbers" to fixed values, updated UI, added new functionality, changed project code strucrute, fixed getters to be getters, changed empty field generation method, changed values to e-num type and hided inside of class, and added description before start, so as total move counter

ArturErmolenko reviewed (approved):
Looks great 🔥 Well done 👍
